### PR TITLE
HITS GPU Correctness Fix and Refactor

### PR DESCRIPTION
**Pull request by jdwapman**
The current implementation of the GPU HITS algorithm appears to be incorrect (see #467 for details).

**Tasks**
- [x] Copy CPU reference algorithm from `dev` branch

- [x] Port the old HITS GPU algorithm to the new API

- [x] Fix the HITS GPU algorithm to give correct results

- [x] Verify correctness vs CPU reference algorithm on small and large datasets

### CODE REVIEW

**Comment by jdwapman** [comment]
Small Datasets Tested:
chesapeake.mtx (undirected)
test_pr.mtx
Tina_AskCog.mtx

File: gunrock/app/hits/hits_enactor.cuh [hunk 1]
```diff
@@ -9,426 +9,463 @@
...
+           if ((this->num_gpus == 1) ||
+                (gpu == this->problem->org_graph->GpT::partition_table[src])) {
+               this -> thread_slices[gpu].init_size = 1;
+               for (int peer_ = 0; peer_ < this -> num_gpus; peer_++) {
+                   auto &frontier = this -> enactor_slices[gpu * this -> num_gpus + peer_].frontier;
+                   frontier.queue_length = (peer_ == 0) ? 1 : 0;
+                   if (peer_ == 0) {
+                       GUARD_CU(frontier.V_Q() -> ForEach(
+                           [src]__host__ __device__ (VertexT &v) {
+                           v = src;
```
Comment (neoblizz) [inline gunrock/app/hits/hits_enactor.cuh]: You can update this from one of the other apps to see how we queue the whole frontier (all nodes) manually.

File: gunrock/app/hits/hits_enactor.cuh [hunk 2]
```diff
@@ -9,426 +9,463 @@
...
+    {
+        cudaError_t retval = cudaSuccess;
+        GUARD_CU(this -> Run_Threads(this));
+        util::PrintMsg("GPU Template Done.", this -> flag & Debug);
+        return retval;
+    }
 };
 
-} // namespace hits
+} // namespace Template
```
Comment (neoblizz) [inline gunrock/app/hits/hits_enactor.cuh]: hits

File: gunrock/app/hits/hits_enactor.cuh (hunk 1 above)
Comment (jdwapman) [reply gunrock/app/hits/hits_enactor.cuh]: I'll look around, but any apps in particular?

File: gunrock/app/hits/hits_app.cu [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    //         fout<< v << "," << h_arank[v] << std::endl;
+    //     }
+    //     fout.close();
+    //     delete[] fout_buf; fout_buf = NULL;
+    //     cpu_timer.Stop();
+    //     parameters.Set("write-time", cpu_timer.ElapsedMillis());
+    // }
+
+    // compute running statistics
+    // <TODO> change NULL to problem specific per-vertex visited marker, e.g. h_distances
```
[the remark this answers is not in the document]
Comment (jdwapman) [reply gunrock/app/hits/hits_app.cu]: Does anything where there are still TODOs need to be changed/optimized?

**Comment by neoblizz** [comment]
I am gonna do a clean run later and then merge this PR.

**Comment by neoblizz** [comment]
Running on large graphs, need help with this. What should I use for max-iter? Or, if you can try it out that will be great. Let the CPU/GPU run for a while in order to make sure that this algorithm works on larger graphs. I will recommend trying delaunay from 10-16 and then maybe trying hollywood-2009. @jdwapman

**Comment by neoblizz** [comment]
```
mosama@bowser:~/forks/jonathan/gunrock/tests/hits (hits_gpu)$ ./bin/test_hits_10.0_x86_64 market /data/gunrock_dataset/large/hollywood-2009/hollywood-2009.mtx --max-iter=1000
Loading Matrix-market coordinate-formatted graph ...
  Reading meta data from /data/gunrock_dataset/large/hollywood-2009/hollywood-2009.mtx.meta
  Reading edge lists from /data/gunrock_dataset/large/hollywood-2009/hollywood-2009.mtx.coo_edge_pairs
  Substracting 1 from node Ids...
  Edge doubleing: 57515616 -> 115031232 edges
  Removed 2279810 duplicate edges and self circles.
  graph loaded as COO in 3.433747s.
Converting 1139905 vertices, 112751422 directed edges ( ordered tuples) to CSR format...Done (0s).
Degree Histogram (1139905 vertices, 112751422 edges):
    Degree 0: 32662 (2.865326 %)
    Degree 2^0: 24445 (2.144477 %)
    Degree 2^1: 50032 (4.389138 %)
    Degree 2^2: 99398 (8.719849 %)
    Degree 2^3: 174673 (15.323470 %)
    Degree 2^4: 231414 (20.301165 %)
    Degree 2^5: 219393 (19.246604 %)
    Degree 2^6: 139586 (12.245406 %)
    Degree 2^7: 78822 (6.914787 %)
    Degree 2^8: 44379 (3.893219 %)
    Degree 2^9: 25905 (2.272558 %)
    Degree 2^10: 13729 (1.204399 %)
    Degree 2^11: 5233 (0.459073 %)
    Degree 2^12: 230 (0.020177 %)
    Degree 2^13: 4 (0.000351 %)

__________________________


--------------------------
 CPU Elapsed: 349179.000000
==============================================
 advance-mode=LB
Using advance mode LB
Using filter mode CULL
__________________________
0        0       0       queue3          oversize :      6839430 ->      112751424
0        0       0       queue3          oversize :      6839430 ->      112751424
--------------------------
Run 0 elapsed: 8670.422077, #iterations = 1000
439048 errors occurred.
CPU Algorithm Results:
Top 10 Ranks:
Hub Ranks:
Vertex ID: 942545, Hub Rank: 0.022894
Vertex ID: 942521, Hub Rank: 0.022885
Vertex ID: 942267, Hub Rank: 0.022870
Vertex ID: 942562, Hub Rank: 0.022793
Vertex ID: 942656, Hub Rank: 0.022754
Vertex ID: 942454, Hub Rank: 0.022698
Vertex ID: 942437, Hub Rank: 0.022628
Vertex ID: 942590, Hub Rank: 0.022605
Vertex ID: 942462, Hub Rank: 0.022590
Vertex ID: 942659, Hub Rank: 0.022564
Authority Ranks:
Vertex ID: 942545, Authority Rank: 0.022894
Vertex ID: 942521, Authority Rank: 0.022885
Vertex ID: 942267, Authority Rank: 0.022870
Vertex ID: 942562, Authority Rank: 0.022793
Vertex ID: 942656, Authority Rank: 0.022754
Vertex ID: 942454, Authority Rank: 0.022698
Vertex ID: 942437, Authority Rank: 0.022628
Vertex ID: 942590, Authority Rank: 0.022605
Vertex ID: 942462, Authority Rank: 0.022590
Vertex ID: 942659, Authority Rank: 0.022564

GPU Algorithm Results:
Top 10 Ranks:
Hub Ranks:
Vertex ID: 942545, Hub Rank: 0.022894
Vertex ID: 942521, Hub Rank: 0.022885
Vertex ID: 942267, Hub Rank: 0.022869
Vertex ID: 942562, Hub Rank: 0.022792
Vertex ID: 942656, Hub Rank: 0.022753
Vertex ID: 942454, Hub Rank: 0.022697
Vertex ID: 942437, Hub Rank: 0.022628
Vertex ID: 942590, Hub Rank: 0.022604
Vertex ID: 942462, Hub Rank: 0.022590
Vertex ID: 942659, Hub Rank: 0.022563
Authority Ranks:
Vertex ID: 942545, Authority Rank: 0.022894
Vertex ID: 942521, Authority Rank: 0.022885
Vertex ID: 942267, Authority Rank: 0.022869
Vertex ID: 942562, Authority Rank: 0.022792
Vertex ID: 942656, Authority Rank: 0.022753
Vertex ID: 942454, Authority Rank: 0.022697
Vertex ID: 942437, Authority Rank: 0.022628
Vertex ID: 942590, Authority Rank: 0.022604
Vertex ID: 942462, Authority Rank: 0.022590
Vertex ID: 942659, Authority Rank: 0.022563
[HITS] finished.
 avg. elapsed: 8670.422077 ms
 iterations: 1000
 min. elapsed: 8670.422077 ms
 max. elapsed: 8670.422077 ms
 load time: 3519.11 ms
 preprocess time: 711.013000 ms
 postprocess time: 464.237928 ms
 total time: 9845.841885 ms
```

**Comment by neoblizz** [comment]
One thing you can note here as well is that this may be a Floating Point comparison. And that is why you don't get some things as `==` even if they are. https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/

**Comment by neoblizz** [comment]
You can also refer to how we do that in the maxflow application: https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/mf/mf_helpers.cuh#L29-L34

**Comment by jdwapman** [comment]
Changed comparison to check scores using tolerance instead of sorted order (set to 1e-6 by default).

**Pull request merged**